File: packages/core-ts/src/relay/session.ts
import { randomBytes, timingSafeEqual } from 'node:crypto'

const SESSION_TTL_MS = 30 * 60 * 1000
const TOKEN_BYTES = 32

export interface ActiveFormSession {
  token: string
  clientId: string
  startedAt: Date
  expiresAt: Date
}

let state: ActiveFormSession | null = null

function now(): Date {
  return new Date()
}

export function claimSession(clientId: string): ActiveFormSession {
  if (state !== null && state.expiresAt > now()) {
    return state
  }
  const startedAt = now()
  state = {
    token: randomBytes(TOKEN_BYTES).toString('base64url'),
    clientId,
    startedAt,
    expiresAt: new Date(startedAt.getTime() + SESSION_TTL_MS)
  }
  return state
}

export function releaseSession(): void {
  state = null
}

export function isSessionActive(): boolean {
  if (state === null) return false
  if (state.expiresAt <= now()) {
    state = null
    return false
  }
  return true
}

export function validateSessionToken(token: string): boolean {
  if (state === null) return false
  if (state.expiresAt <= now()) {
    state = null
    return false
  }
  const a = Buffer.from(state.token)
  const b = Buffer.from(token)
  const isLengthEqual = a.length === b.length
  const compareB = isLengthEqual ? b : a
  return timingSafeEqual(a, compareB) && isLengthEqual
}

export function getActiveSession(): ActiveFormSession | null {
  if (state === null) return null
  if (state.expiresAt <= now()) {
    state = null
    return null
  }
  return state
}

File: packages/core-py/src/mcp_core/relay/session.py
"""Single active form session token state — D11.

Prevents racing 2+ Claude Code sessions concurrently filling the relay form
for the same server. Only one bridge holds the active form session at a time;
subsequent claim_session() calls return the existing session info without
issuing a new token.
"""

from __future__ import annotations

import secrets
from mcp_core.crypto.timing import timing_safe_equal
import threading
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Optional

SESSION_TTL = timedelta(minutes=30)
TOKEN_NBYTES = 32


@dataclass
class ActiveFormSession:
    token: str
    client_id: str
    started_at: datetime
    expires_at: datetime


_state: Optional[ActiveFormSession] = None
_lock = threading.Lock()


def _now() -> datetime:
    return datetime.now(timezone.utc)


def claim_session(client_id: str) -> ActiveFormSession:
    """Claim form session. If active and unexpired, return existing.
    Otherwise mint a new session bound to client_id."""
    global _state
    with _lock:
        if _state is not None and _state.expires_at > _now():
            return _state
        now = _now()
        _state = ActiveFormSession(
            token=secrets.token_urlsafe(TOKEN_NBYTES),
            client_id=client_id,
            started_at=now,
            expires_at=now + SESSION_TTL,
        )
        return _state


def release_session() -> None:
    """Clear active session (called after submit success or expire)."""
    global _state
    with _lock:
        _state = None


def is_session_active() -> bool:
    """Return True if there's an active unexpired session."""
    global _state
    with _lock:
        if _state is None:
            return False
        if _state.expires_at <= _now():
            _state = None
            return False
        return True


def validate_session_token(token: str) -> bool:
    """Validate token against active session. Returns False if no active or mismatch."""
    global _state
    with _lock:
        if _state is None:
            return False
        if _state.expires_at <= _now():
            _state = None
            return False
        return timing_safe_equal(_state.token, token)


def get_active_session() -> Optional[ActiveFormSession]:
    """Return active session info or None. Used by config__open_relay tool."""
    global _state
    with _lock:
        if _state is None:
            return None
        if _state.expires_at <= _now():
            _state = None
            return None
        return _state
